### PR TITLE
Support intersection of versions

### PR DESCRIPTION
**Pull request by waj**
This adds support for version specifiers like: `>= 1.0.0, < 2.0`
I need this to implement #365, but I think it's also nice addition. Other dependency managers also support expressions like that.

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -167,8 +167,9 @@ A version requirement (String).
 - It may be a version number.
 - It may be `"*"` if any version will do.
 - The version number may be prefixed by an operator: `<`, `<=`, `>`, `>=` or `~>`.
+- Many expressions can be separated by commas.
```
Comment (straight-shoota) [inline SPEC.md:170]:
```suggestion
- Multiple requirements can be separated by commas.
```

**Comment by Sija** [comment]
@waj FWIW, I believe that a _compound_ keyword is more apt in this case than _intersection_. I'd use _compound requirement_ - as in bundler, see:

https://github.com/rubygems/rubygems/blob/6f1037c67557a33a81ead0a6b2b5867d601f0d08/bundler/spec/install/gems/resolving_spec.rb#L241

**Comment by Blacksmoke16** [comment]
Compose calls them [version ranges](https://getcomposer.org/doc/articles/versions.md#version-range).

They also seem to allow `||` together ranges.  `>=1.0 <1.1 || >=1.2`.

**Comment by Sija** [comment]
@Blacksmoke16 I don't know about the Composer, but they're not _ranges_, that's for sure :)

**Comment by Blacksmoke16** [comment]
@Sija How is it not? `>= 1.0.0, < 2.0` would be `1.0.0...2.0.0`, i.e. any version greater than or equal to `1.0.0` but less than `2.0.0`.

**Comment by Sija** [comment]
@Blacksmoke16 because `>= 1.0.0, < 2.0, != 1.1`

**Comment by Blacksmoke16** [comment]
@Sija Still could be interpreted as `1.0.0...1.1.0 || 1.2.0...2.0.0`. 🤷.

**Comment by Sija** [comment]
@Blacksmoke16 Yeah, that's a _compound_ (range) statement you've got there...

**Comment by waj** [comment]
I prefer the word "intersection" because it's more explicit about how the expressions are composed. In the future, if we add other operators like `||` or `...` the text can be rephrased.

**Comment by Sija** [comment]
@waj That's not _intersection_ though... And the comma used there is analogous to `&&` (and) logical operator.

**Comment by waj** [comment]
@Sija It's the intersection of the version sets defined by each expression. In math intersection and logical and are very close concepts. Also, I'm not the only one using this term for this: https://dart.dev/tools/pub/dependencies#traditional-syntax

**Comment by Sija** [comment]
@waj Not if you add `!=` operator to the mix.

**Comment by bcardiff** [comment]
@Sija It’s intersection. All comma separated constraints need to be satisfied. Even if != is supported. https://www.siyavula.com/read/maths/grade-10/probability/14-probability-03

If or/union is added later is another matter.

**Comment by Sija** [comment]
> ### Intersection
> The intersection of two sets is a new set that contains all of the elements that are in both sets. The intersection is written as 𝐴∩𝐵 or “𝐴 and 𝐵”.

Quite likely I'm misunderstanding the term in this context but I don't see any (_intersecting_) _sets_ here. I mean we have constraints, yes, but not sets.

By set intersection I'd understand having `1.0-1.2, 1.1-1.3`. OTOH `>= 1.0.0, < 2.0` defines a _set_ of versions between `1.0` and `2.0`, which is closer to a range in this case than to a _intersection_... adding multiple closes turns that into a _compound_ statement.

But maybe my brain is playing tricks on me 🤷‍♂️

**Comment by waj** [comment]
@Sija do you realise we're discussing for a word it doesn't even appear in the documentation, right? It's just the issue name and a spec description at most 😄 

It's not brain tricks. Just different ways to see the same. When you say `>= 1.0.0` that's a range too. Just an unbounded range. And a range is a set. And sets can be intersected to create a new set. But while all ranges are sets, not all sets are ranges. For example, `>= 1.0.0, < 2.0, != 1.2.3` cannot be expressed as a single range (note that `!=` is currently not supported). Still can be interpreted as the intersection of three sets. Or a union of two ranges if you want, but for me it's harder to think in ranges because is more complicated to generalize.

**Comment by Sija** [comment]
With this merged in, it would be good to add support for `!=` operator.

**Pull request merged**